Simplified Chinese Translation, working with TexLive.net

libralibra opened the pull request:
1. Add Simplified Chinese Translation by following translation instructions
<del>2. Add a temporary patch to make all examples work with TexLive.net (click the TexLive.net button to get the examples that contains Chinese punctuations compiled successfully)</del>

3. Rewrite `language-01.md` to reflect Chinese language-specific issues
4. Check all examples in the forked git-page site before bringing back `CNAME.txt` (https://libralibra.github.io/learnlatex.github.io/zh/)

**UPDATE**:
4. Update the default engine from `xelatex` to `lualatex`
5. Remove the unnecessary reference to `xeCJK` package

davidcarlisle commented on zh/lesson-10.md:
```diff
@@ -24,8 +24,19 @@ toc-description: "数学模式和数学符号。"
-\usepackage[T1]{fontenc}
+% !TEX program=xelatex
+
+% 临时patch，否则使用中文标点，TexLive.net会编译错误
```
is there an issue on texlive.net that we could fix there rather than have a patch in the document?  Ssorry I can not read the comment well enough, something about chinese punctuation not working at texlive.net? It shoudl be a standard texlive 2024 installed there, last updated a few days ago

davidcarlisle replied:
Oh sorry, I see, you were needing the latest version of the L3 layer

I just updated texlive on the server ans it works without this patch now

https://texlive.net/run?%25%20!TEX%20program=xelatex%0A%0A%5Cdocumentclass%5BUTF8%5D%7Bctexart%7D%0A%5Cusepackage%7BxeCJK%7D%0A%5Cbegin%7Bdocument%7D%0A%E4%B8%80%E4%B8%AA%E5%8C%85%E5%90%AB%E8%A1%8C%E5%86%85%E6%95%B0%E5%AD%A6%E5%85%AC%E5%BC%8F%E7%9A%84%E5%8F%A5%E5%AD%90%EF%BC%9A%24y%20=%20mx%20+%20c%24%E3%80%82%0A%E5%8F%A6%E4%B8%80%E4%B8%AA%E5%8C%85%E5%90%AB%E8%A1%8C%E5%86%85%E6%95%B0%E5%AD%A6%E5%85%AC%E5%BC%8F%E7%9A%84%E5%8F%A5%E5%AD%90%EF%BC%9A%245%5E%7B2%7D=3%5E%7B2%7D+4%5E%7B2%7D%24%E3%80%82%0A%0A%E7%AC%AC%E4%BA%8C%E6%AE%B5%E5%8C%85%E5%90%AB%E6%98%BE%E7%A4%BA%E6%95%B0%E5%AD%A6%E5%85%AC%E5%BC%8F%E3%80%82%0A%5C%5B%0A%20%20y%20=%20mx%20+%20c%0A%5C%5D%0A%E7%9C%8B%E7%9C%8B%E6%AE%B5%E8%90%BD%E5%A6%82%E4%BD%95%E5%9C%A8%E6%98%BE%E7%A4%BA%E5%85%AC%E5%BC%8F%E5%90%8E%E7%BB%A7%E7%BB%AD%E3%80%82%0A%5Cend%7Bdocument%7D

libralibra replied:
thanks, I will remove all the patches and make another test.

davidcarlisle commented:
Thank you for this.

It would be really good to have more translations of the site showing non-latin scripts.

As you may know, a translation was started some time ago at 
https://github.com/CTeX-org/learnlatex.github.io
but never completed.

@libralibra  I hope you don't mind if I ask if  @stone-zeng or @LogCreative  could give a quick review of this PR, as we can not read the text it is good to have another reader look at it if possible.

I left one inline comment asking about "patches for texlive.net"

josephwright commented:
I'm a bit worried about the need for the patch here, not because we don't want patched files but because it's only needed for XeTeX. Like @davidcarlisle, I can't read the text but can see in the examples you've used `xeCJK` and non-standard classes. I'm not in a position to comment on the latter (do the standard classes not work at all with Chinese?), but in terms of support I'd like to see us move toward LuaTeX + `babel`, here I suspect with `luatex-ja`. Perhaps we are not yet in a position to do that in the lessons, but we should mark something down to review - not just this PR, but the site generally.

josephwright commented:
That said, I'm very pleased to see a PR for Chinese, and it's not meant to mean I don't want to merge - more a 'next development step' comment.

lvjr commented:
When you are using `ctexart` you need not to load `xeCJK`, because it will check tex engine and automatically load `CJK` for pdftex, `xeCJK` for xetex and `luatex-ja` for luatex. Simple `ctexart` documents should work on all engines.

davidcarlisle commented:
> When you are using `ctexart` you need not to load `xeCJK`, because it will check tex engine and automatically load `CJK` for pdftex, `xeCJK` for xetex and `luatex-ja` for luatex. Simple `ctexart` documents should work on all engines.

That is good to know, it would be good if it said that somewhere (in Chinese:-) on this site.

Also perhaps the basic help

https://libralibra.github.io/learnlatex.github.io/zh/help#%E7%A4%BA%E4%BE%8B

could duplicate the initial example and have a `%!TeX lualatex`  version as well given the general issues with xelatex support as described at https://www.texdev.net/2024/11/05/engine-news-from-the-latex-project or the latest latex news with the November release.

libralibra commented:
> Thank you for this.
> 
> It would be really good to have more translations of the site showing non-latin scripts.
> 
> As you may know, a translation was started some time ago at https://github.com/CTeX-org/learnlatex.github.io but never completed.
> 
> @libralibra I hope you don't mind if I ask if @stone-zeng or @LogCreative could give a quick review of this PR, as we can not read the text it is good to have another reader look at it if possible.
> 
> I left one inline comment asking about "patches for texlive.net"

Thank you for your reply. I also came across that forked repository and wondered why there wasn't a Chinese translation of this excellent LaTeX primer. Unfortunately, it appears to have been abandoned before completion, so I created a new fork. I may add a Traditional Chinese translation soon as well.

libralibra commented:
> > When you are using `ctexart` you need not to load `xeCJK`, because it will check tex engine and automatically load `CJK` for pdftex, `xeCJK` for xetex and `luatex-ja` for luatex. Simple `ctexart` documents should work on all engines.
> 
> That is good to know, it would be good if it said that somewhere (in Chinese:-) on this site.
> 
> Also perhaps the basic help
> 
> https://libralibra.github.io/learnlatex.github.io/zh/help#%E7%A4%BA%E4%BE%8B
> 
> could duplicate the initial example and have a `%!TeX lualatex` version as well given the general issues with xelatex support as described at https://www.texdev.net/2024/11/05/engine-news-from-the-latex-project or the latest latex news with the November release.

That's good to know. I have tested in my forked repo and it did work. I will update the examples accordingly to remove the use of xeCJK and change the default engine to LuaLaTeX as well. 👍

libralibra commented:
> I'm a bit worried about the need for the patch here, not because we don't want patched files but because it's only needed for XeTeX. Like @davidcarlisle, I can't read the text but can see in the examples you've used `xeCJK` and non-standard classes. I'm not in a position to comment on the latter (do the standard classes not work at all with Chinese?), but in terms of support I'd like to see us move toward LuaTeX + `babel`, here I suspect with `luatex-ja`. Perhaps we are not yet in a position to do that in the lessons, but we should mark something down to review - not just this PR, but the site generally.

Thank you for your reply. I’ve read through all the helpful comments and conducted some tests and experiments.

It seems that LuaLaTeX is indeed the recommended engine, and it’s actually faster than XeLaTeX when compiling examples on texlive.net. I’ve updated the default engine in the Chinese examples to LuaLaTeX to reflect this.

Regarding the non-standard document classes, ctex-* classes are considered best practice for Chinese-language documents, as they are specifically designed to handle Chinese characters and punctuation effectively.

I'll close this PR and open a new one with these changes. This has been a valuable learning experience, and I look forward to adapting these insights into my own LaTeX book projects as well.

libralibra commented:
> When you are using `ctexart` you need not to load `xeCJK`, because it will check tex engine and automatically load `CJK` for pdftex, `xeCJK` for xetex and `luatex-ja` for luatex. Simple `ctexart` documents should work on all engines.

Thank you very much. Got it now :-)

davidcarlisle commented:
@libralibra 
>I'll close this PR and open a new one with these changes. 

There is no need to open a new PR simply push to your fork and this PR will automatically update, then it is easier to see what changes have been made.

Also don't worry too much about the CNAME file, if you take `learnlatex.org` out of your copy so that you do not "steal" the DNS name, then it is true that directly merging your PR would change that here as well, but we know that happens and can fix it while merging the code so you do not have to keep changing CNAME depending on whether you are testing locally or making version for us to merge.

u-fischer commented:
Why do many of the examples use the `UTF8` class option? It seems to compile fine without it.

davidcarlisle commented:
> Why do many of the examples use the `UTF8` class option? It seems to compile fine without it.

I think UTF8 is the default now (and the only supported encoding for luatex or xetex), so this could be dropped, (via google translate) the ctex manual says:

**4.2 Chinese encoding**

**GBK**
indicates the encoding used when writing the document. The CTEX macro set cannot detect the actual encoding format of the document source file, so the user needs to declare it through the option

**UTF8**
If not explicitly specified, UTF-8 encoding is used by default.
Updated: 2019-11-10
When compiled with XƎLATEX, LuaLATEX or upLATEX, the CTEX macro set is forced to use UTF-8 encoding, and the GBK option
is invalid at this time; when compiled with (pdf)LATEX, the CTEX macro set uses UTF-8 encoding by default, but users can also explicitly declare the GBK
option to make the CTEX macro set process documents in GBK encoding.
Users need to ensure that the compilation method, source file encoding, and macro package encoding options are consistent.
We recommend that you always use UTF-8 encoding when writing new documents, and only leave GBK encoding for legacy documents.

libralibra commented:
> Why do many of the examples use the `UTF8` class option? It seems to compile fine without it.

It's a personal practice of 'explicit is better than implicit' from the Zen of Python. But I am happy to remove it as the examples work fine without it as you suggested.

libralibra commented:
> > Why do many of the examples use the `UTF8` class option? It seems to compile fine without it.
> 
> I think UTF8 is the default now (and the only supported encoding for luatex or xetex), so this could be dropped, (via google translate) the ctex manual says:
> 
> **4.2 Chinese encoding**
> 
> **GBK** indicates the encoding used when writing the document. The CTEX macro set cannot detect the actual encoding format of the document source file, so the user needs to declare it through the option
> 
> **UTF8** If not explicitly specified, UTF-8 encoding is used by default. Updated: 2019-11-10 When compiled with XƎLATEX, LuaLATEX or upLATEX, the CTEX macro set is forced to use UTF-8 encoding, and the GBK option is invalid at this time; when compiled with (pdf)LATEX, the CTEX macro set uses UTF-8 encoding by default, but users can also explicitly declare the GBK option to make the CTEX macro set process documents in GBK encoding. Users need to ensure that the compilation method, source file encoding, and macro package encoding options are consistent. We recommend that you always use UTF-8 encoding when writing new documents, and only leave GBK encoding for legacy documents.

Thank you for quoting the documents. All examples work fine without the explicit UTF8 definition indeed.

muzimuzhi commented:
To better review translations, is it possible to first have a clean copy of English version under `./zh`, which is (temporarily) hidden on learnlatex.github.io? That would be basically merging the first commit (https://github.com/learnlatex/learnlatex.github.io/pull/221/commits/affb6e8360811fd6461e786d1dbfb37a3f9de272) in this PR first.

Reviewing selected commits does the trick, but the English version would still be missing in contexts of review comments.

Some general questions:
- To provide both Simplified and Traditional Chinese translations in the future, need the directory name for Simplified Chinese be more specific than `./zh`, e.g., `zh-Hans` or even `zh-Hans-CN`?
- I recommend inserting explicit space characters between Chinese and non-Chinese characters in `.md` files. For example "在Overleaf中打开" (Open in Overleaf) -> "在 Overleaf 中打开". (I can add them when reviewing, just ask for agreement.)
- I don't think "您" (your honor) is commonly used in Simplified Chinese learning resources. Just use "你"?

davidcarlisle commented:
@muzimuzhi  I'm not sure I did quite what you intended

I locally cherry picked affb6e8360811fd6461e786d1dbfb37a3f9de272 into main (which added the English text under zh as expected) then pushed here and that shows in the three dot diff as in

https://github.com/learnlatex/learnlatex.github.io/compare/main...libralibra:learnlatex.github.io:mainch

But the PR still shows the difs against which it is forked

muzimuzhi commented:
@davidcarlisle Thank you! According to how three-dot Git diff works unfortunately this is expected behavior. 

> The three-dot comparison shows the difference between the latest common commit of both branches (merge base) and the most recent version of the topic branch.
> [Three-dot Git diff comparison](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#three-dot-git-diff-comparison) | GitHub Docs

Now it shows "This branch has conflicts that must be resolved". Not sure if it's easy to add a new commit which reverts https://github.com/learnlatex/learnlatex.github.io/commit/affb6e8360811fd6461e786d1dbfb37a3f9de272 to the head branch. I will try it locally.

davidcarlisle commented:
@muzimuzhi  @libralibra  there is some discussion with @stone-zeng in #77 and I think the suggestion there was to use zh-hans (part of the discussion relates to earlier versions of the site when the language switching may not have been quite the same)

davidcarlisle commented:
I have checked that the language switching does work using the longer codes such as `zh-hans` . There are some comments in the earlier issue implying 2-letter codes are assumed, but it looks like I fixed that at the time.  As far as I can see, language selection works on my fork at

https://davidcarlisle.github.io/learnlatex.github.io/zh-hans/

so if you git rename the directory to zh and adjust the `lang:` and `permalink` entries to use that, it should all work.

libralibra commented:
> I have checked that the language switching does work using the longer codes such as `zh-hans` . There are some comments in the earlier issue implying 2-letter codes are assumed, but it looks like I fixed that at the time. As far as I can see, language selection works on my fork at
> 
> https://davidcarlisle.github.io/learnlatex.github.io/zh-hans/
> 
> so if you git rename the directory to zh and adjust the `lang:` and `permalink` entries to use that, it should all work.

Thanks for your reply. Yes, the longer language code did work.

I have made changes slowly to reflect what has been discussed above. 

1. updated translation in Simplified Chinese (I may start working on the Traditional Chinese translation after this) using the longer language code `zh-hans`
2. add extra space before and after English words in the Chinese translation.
3. there's no need to keep a copy of the `en` folder in translated directory if I read the `TRANSLATIONS.md` correctly

davidcarlisle commented:
@libralibra  thanks for this

> 
> I have made changes slowly to reflect what has been discussed above, as shown in the test site:>
>
> https://libralibra.github.io/learnlatex.github.io/zh-hans/
> 

Thanks for this. If you let me know when you think it is done, I will merge.

If other  changes come up after the merge you can always make a new PR to add corrections of course.

@muzimuzhi  @stone-zeng similarly comments before or after the merge are always welcome.

> 
> 1.    updated translation in Simplified Chinese (I may start working on the Traditional Chinese translation after this) using the longer language code zh-hans

Thank you. It is simpler to limit a PR to one script, so if you could make a new PR for any zh-hant additions rather than adding both here that would be helpful . Thank you.

> 2.    add extra space before and after English words in the Chinese translation.

thanks

> 3.   there's no need to keep a copy of the en folder in translated directory if I read the TRANSLATIONS.md correctly

Yes there should be no english under the zh-hans directory. What @muzimuzhi  meant by the above comment about having english (which you could do for any new zh-hant translation). Is to start by populating `zh-hant` from a copy of the `en` directory rather than empty stubs, and then the files changed diff view in any PR will show the English (as `-` removed) and Chinese as (`+` added) on the same page, for easier proof reading.

davidcarlisle commented:
@libralibra  also don't worry too much about the conflict under `zh` we can fix that locally if not fixed at your side.

libralibra commented:
> @libralibra thanks for this
> 
> > I have made changes slowly to reflect what has been discussed above, as shown in the test site:>
> > https://libralibra.github.io/learnlatex.github.io/zh-hans/
> 
> Thanks for this. If you let me know when you think it is done, I will merge.
> 
> If other changes come up after the merge you can always make a new PR to add corrections of course.
> 
> @muzimuzhi @stone-zeng similarly comments before or after the merge are always welcome.
> 
> > 1. updated translation in Simplified Chinese (I may start working on the Traditional Chinese translation after this) using the longer language code zh-hans
> 
> Thank you. It is simpler to limit a PR to one script, so if you could make a new PR for any zh-hant additions rather than adding both here that would be helpful . Thank you.
> 
> > 2. add extra space before and after English words in the Chinese translation.
> 
> thanks
> 
> > 3. there's no need to keep a copy of the en folder in translated directory if I read the TRANSLATIONS.md correctly
> 
> Yes there should be no english under the zh-hans directory. What @muzimuzhi meant by the above comment about having english (which you could do for any new zh-hant translation). Is to start by populating `zh-hant` from a copy of the `en` directory rather than empty stubs, and then the files changed diff view in any PR will show the English (as `-` removed) and Chinese as (`+` added) on the same page, for easier proof reading.

Thank you for your reply. The current state of the Simplified translation can be considered as finished.
I've removed zh-hant from the yaml files as I felt the same about making a new PR for the Traditional translation would be more reasonable.

davidcarlisle commented:
@libralibra  thanks a lot

https://www.learnlatex.org/zh-hans/

I hope you can read that better than I can:-)

davidcarlisle commented:
@libralibra  I  added a `{: .norun :}` in the "template" example in section 5 to stop it adding the buttons to try running `\documentclass{<meta-styntax-for-class-name>}`

3ba303c2ba05bc93aa49b6fa38c6976ed958e90e

davidcarlisle commented:
@libralibra  should there be a translation for the "meta" heading in the page footer?

![image](https://github.com/user-attachments/assets/83ae2448-778f-4697-9802-908a494b4f8c)


It seems most of the the languages failed to translate this (or translated it unchanged to Meta)  but it looks a bit odd in latin script here (because I can read it:-) It would need to be in `_data/translations.yaml` but you could just supply a text here rather than make a new PR just for one word (or if you think it should stay as it is, that is OK too)

Also I notice that while generally we force `Noto Sans` Font, for Chinese it is falling through to whatever font is supplied on the local system (I get `Microsoft YaHei`)  We could get `Noto Sans SC` as a web font eg from google fonts but that does force a fairly large font download on users who may? prefer to use their existing system Chinese font, especially on mobile.

libralibra commented:
> @libralibra should there be a translation for the "meta" heading in the page footer?
> 
> ![image](https://private-user-images.githubusercontent.com/1268738/390892625-83ae2448-778f-4697-9802-908a494b4f8c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI5ODM5NzYsIm5iZiI6MTczMjk4MzY3NiwicGF0aCI6Ii8xMjY4NzM4LzM5MDg5MjYyNS04M2FlMjQ0OC03NzhmLTQ2OTctOTgwMi05MDhhNDk0YjRmOGMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMzBUMTYyMTE2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTE2ODY4MmNmMGQ3MTFhYjcwMGUzNDFhYmYzMzdlZTI4NDc1YjI1ODVjODU2NTA1YmJlZTE4YWY5ZTk0NTM4MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.jOuRa8ZdDEmuA05fqU5HsFN24IUgAyEb96qWABd6Afk)
> 
> It seems most of the the languages failed to translate this (or translated it unchanged to Meta) but it looks a bit odd in latin script here (because I can read it:-) It would need to be in `_data/translations.yaml` but you could just supply a text here rather than make a new PR just for one word (or if you think it should stay as it is, that is OK too)
> 
> Also I notice that while generally we force `Noto Sans` Font, for Chinese it is falling through to whatever font is supplied on the local system (I get `Microsoft YaHei`) We could get `Noto Sans SC` as a web font eg from google fonts but that does force a fairly large font download on users who may? prefer to use their existing system Chinese font, especially on mobile.

@davidcarlisle 
It may be translated to 元数据, but it might not be easy to find a proper word in Chinese to precisely act as Meta in English. 

I struggled a little bit between using the translation and keeping the original word during the translation process. You could do either I think. For a person who would like to learn LaTeX, it shouldn't be a problem.

Regarding the font, it should be fine to let the local system selects whatever it likes as it won't affect the user experience too much.

davidcarlisle commented:
@libralibra  thanks I think I'll update the text, "Meta" is not really that good a heading in English (and is probably difficult to translate to many languages) I don't think we need to chaneg the english but I don't think it's important to preserver the meaning and highlight that heading by being the only Latin script heading.  pushed 3f127360

I'll leave the font as it is.
